fixup! CIP-0021: Restrictions on transactions signed by hardware wallets (#4)
vacuumlabs/CIPs: CIP-0021/CIP-0021.md

The numbers of inputs, outputs, certificates, withdrawals, tokens in an asset group, and other similar transaction elements must not exceed `UINT16_MAX`, i.e. 65535.

**Optional empty lists and maps**

Unless mentioned otherwise in this CIP, optional empty lists and maps must not be included as part of the transaction body or its elements.

**Outputs**

> Outputs containing no multiassets must be serialized as a simple tuple. I.e. `[address, coin]` instead of the alternative `[address, [coin, {}]]`.

Reviewer: ```suggestion
Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin]` instead of `[address, [coin, {}]]`.
```

Reviewer: It seems to me we support empty asset groups --- should it mentioned in this document? Perhaps not, but it's a natural question to ask and it's not obvious if it falls under "optional empty lists and maps".

Author: Trezor doesn't allow empty asset groups. I see nothing restricting this on Ledger. Tokens are shown on Trezor only as a combination of policyId+assetName, thus I don't know what should be shown if an empty asset group is provided - is it shown on Ledger? If so how? I might check myself tomorrow.

To clarify the CIP though, we could add "If a map's key is also an empty list or an empty map it is also not to be included unless stated otherwise." to "Optional empty lists and maps". What do you think?

Author: Hm, it seems to me that Ledger actually crashes (restarts) when signing a transaction with the following output with empty `tokens` in `tokenBundle`:
```
{
  destination: {
    type: TxOutputDestinationType.THIRD_PARTY,
    params: {
      addressHex:
        "82d818582183581c9e1c71de652ec8b85fec296f0685ca3988781c94a2e1a5d89d92f45fa0001a0d0c2561",
    },
  },
  amount: "3003112",
  tokenBundle: [
    {
      policyIdHex:
        "95a292ffee938be03e9bae5657982a74e9014eb4960108c9e23a5b39",
      tokens: [],
    },
  ],
},
```
When I remove the `tokenBundle` or add tokens to `tokens` it works.

Author: I just realised we might be mixing up `tokenBundle` and `asset group` - I'm not sure which is which right now. Sending an empty `tokenBundle: []` produces the same tx hash as not sending the `tokenBundle` does.

Since withdrawals are represented as a map of reward accounts, withdrawals also need to be sorted in accordance with the specified canonical CBOR format. A transaction must not contain duplicate withdrawals. All withdrawals included in a transaction must have the same type of stake credential i.e. either 0 - `addr_keyhash` or 1 - `scripthash`. The stake credential type must also be consistent with the type used for certificates.

**Auxiliary data**

> Auxiliary must be encoded in their "tuple" format i.e.:

Reviewer: ```suggestion
HW wallets do not serialize auxiliary data because of their complex structure --- they only include the given auxiliary data hash in the transaction body. The only exception is Catalyst voting registration because it requires a signature computed by the hw wallet. 

In this exceptional case, auxiliary data must be encoded in their "tuple" format:
```

Author: https://github.com/vacuumlabs/CIPs/pull/4/commits/707b4f37e97c2b4287930bc49efa53ec864c9ac7

```
[ transaction_metadata: { * transaction_metadatum_label => transaction_metadatum }, auxiliary_scripts: [ * native_script ]]
```

instead of:

```
{ * transaction_metadatum_label => transaction_metadatum }

> ```

Reviewer: Let's replace this "instead of" part with
"The item `auxiliary_scripts` must be an array of length 0."

Author: https://github.com/vacuumlabs/CIPs/pull/4/commits/707b4f37e97c2b4287930bc49efa53ec864c9ac7

Similarly to multiassets, allowing duplicate withdrawals might lead to inconsistencies between what is displayed to the user and how nodes and other tools might interpret the duplicate keys.

**Auxiliary data**

> The specified auxiliary data format is used in order to be compatible with other Cardano tools, which mostly use this serialization format.

Reviewer: ```suggestion
The specified auxiliary data format was chosen in order to be compatible with other Cardano tools, which mostly use this serialization format.
```

Author: https://github.com/vacuumlabs/CIPs/pull/4/commits/707b4f37e97c2b4287930bc49efa53ec864c9ac7
